app/controllers/pieces_controller.rb

```diff
@@ -101,7 +101,6 @@ def game_has_two_players?(game)
 
   def check_test(piece, x, y)
     return false if piece.can_take?(helpers.get_piece(x, y, piece.game)) && piece.type == 'King'
-    return false if piece.can_take?(helpers.get_piece(x, y, piece.game)) && !piece.puts_self_in_check?(x, y)
 
     if piece.puts_self_in_check?(x, y)
       flash.now[:alert] << 'You cannot put/leave yourself in Check.'
```

Comment:
Okay can't comment any further down then this because no changes were committed further down but as I look at this if statement this seems like it should be the first check in the check_test method. (or better yet moved to the `valid_move?` method) Having this further down means someone could leave themselves open to ending the turn in check. Which leads to all sorts of problems.

Reply:
The order for the first two lines doesn't matter. One is checking to see if a king piece can get itself out of check and the other is to ensure that you are not putting yourself into check when moving a piece. 

I think this is fine where it is, as it is a test to see if you would be put in check, a check test. 

This line `return false unless piece.puts_enemy_in_check?(x, y)` needs to stay towards the bottom as we do not want the method to return false if a move doesn't always put the enemy in check. 

I'm not really sure how having this line where it is could lead to any problems. If you notice any let me know and we can fix it.

Reply:
Also `valid_move?` is different for each piece model. We would have to add this functionality to every model. I think it is better sitting inside one method.

Reply:
I would agree with that. You could move inside is obstructed since every piece calls that method. (except for King so 2 vs 1 and that means it is still better here).

Reply:
>  One is checking to see if a king piece can get itself out of check

I do not think line 103 checks if a king can move out of check. It checks whether the piece at the destination tile is of the opposite color allowing it to be take and if the piece being moved is a king.

Reply:
> This line return false unless piece.puts_enemy_in_check?(x, y) needs to stay towards the bottom as we do not want the method to return false if a move doesn't always put the enemy in check.

This is not true since the first thing we want to do in the `check_test` method is see if it actually puts the opposing player in check. Also all the lines above this also return false so if we wanted this to return true before false it would need to be moved down. But again we want this method to return false first thing for the many moves that do not actually put the opposing player in check.

Reply:
Yeap, you are right there about line 103. I'm not even sure if it's necessary now. Because:
```
if piece.puts_self_in_check?(x, y)
      flash.now[:alert] << 'You cannot put/leave yourself in Check.'
      return false
    end
```
would cover the same thing. It doesn't really matter if a piece is a king or not here.

Reply:
And sorry saw the change to late. GitHub does not use ActionCable I guess so did not see change till after I posted these comments.